Fix Check For num_entries for CpuId #119

gabhijit opened the pull request:
Added a check to verify that the number of entires passed is less than
or equal to `KVM_MAX_CPUID_ENTRIES` (`Kvm.get_cpuid`,=`VcpuFd.get_cpuid2`).

Added test cases for failure scenarios for `get_emulated_cpuid` and
`get_supported_cpuid`.

Renamed param `max_entries_count` to `num_entries` consistent with
`VcpuFd.get_cpuid2`

gabhijit committed:
Fix Check For num_entries for CpuId #119

Added a check to verify that the number of entires passed is less than
or equal to `KVM_MAX_CPUID_ENTRIES` (`Kvm.get_cpuid`,
`VcpuFd.get_cpuid2`).

Added test cases for failure scenarios for `get_emulated_cpuid` and
`get_supported_cpuid`.

Renamed param `max_entries_count` to `num_entries` consistent with
`VcpuFd.get_cpuid2`

Signed-off-by: Abhijit Gadgil <gabhijit@iitbombay.org>

gabhijit commented:
Odd that I do not get the same warnings if I do a local `cargo build` and or `cargo test`? Not sure why those warnings. 

Does commit format require a 'signed-off-by'? line May be that's causing the failure?

acatangiu commented on src/ioctls/system.rs:
```diff
@@ -228,13 +228,18 @@ impl Kvm {
+    fn get_cpuid(&self, kind: u64, num_entries: usize) -> Result<CpuId> {
+        if num_entries > KVM_MAX_CPUID_ENTRIES {
+            // Returns the same error the underlying `ioctl` would have sent.
+            return Err(errno::Error::new(ENOMEM));
```
nit: Actually newer kernels return `E2BIG`, you could do the same.

gabhijit replied:
Actually `E2BIG` is returned if the passed number of entries is too low and `ENOMEM` is returned if the passed number of entries is too high. So I believe in this case `ENOMEM` is right - as per [this documentation](https://www.kernel.org/doc/html/latest/virt/kvm/api.html?highlight=kvm%20ioctl)

acatangiu commented on src/ioctls/system.rs:
```diff
@@ -595,6 +600,14 @@ mod tests {
 
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    #[test]
+    fn test_get_supported_cpuid_fail_nentries_high() {
```
nit: this is small enough to be another check in the `test_get_supported_cpuid()` test instead of a standalone test

gabhijit replied:
Okay will do it there.

acatangiu commented on src/ioctls/system.rs:
```diff
@@ -605,6 +618,14 @@ mod tests {
 
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    #[test]
+    fn test_get_emulated_cpuid_fail_nentries_high() {
```
nit: this is small enough to be another check in the `test_get_emulated_cpuid()` test instead of a standalone test

gabhijit replied:
Okay will do it there.

acatangiu commented on src/ioctls/vcpu.rs:
```diff
@@ -376,6 +376,11 @@ impl VcpuFd {
     pub fn get_cpuid2(&self, num_entries: usize) -> Result<CpuId> {
+        if num_entries > KVM_MAX_CPUID_ENTRIES {
+            // Returns the same error the underlying `ioctl` would have sent.
+            return Err(errno::Error::new(ENOMEM));
```
nit: same here, `E2BIG` is better

gabhijit replied:
Same as above. I think `ENOMEM` is correct.

acatangiu reviewed (commented):
The kernel already does all the validations included in this PR. The overhead here is minimal but also not necessary. I'll leave the final decision to the other maintainers.

andreeaflorescu commented on src/ioctls/system.rs:
```diff
@@ -267,8 +272,8 @@ impl Kvm {
     #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
-    pub fn get_emulated_cpuid(&self, max_entries_count: usize) -> Result<CpuId> {
-        self.get_cpuid(KVM_GET_EMULATED_CPUID(), max_entries_count)
+    pub fn get_emulated_cpuid(&self, num_entries: usize) -> Result<CpuId> {
```
I think we should also update the documentation of the function to specify that it returns an error when `num_entries` > KVM_MAX_CPUID_ENTRIES. Same for `get_supported_cpuid`.

gabhijit committed:
documentation: Added returned Error value #119

Changed the documentation of `get_{supported|emulated}_cpuid` to
describe returned error.

Also, fixed a copy-paster error in he documentation for
`get_supported_cpuid` (Example was calling `get_emulated_cpuid`)

Signed-off-by: Abhijit Gadgil <gabhijit@iitbombay.org>
